apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/bytebuddy/CacheableTransformerDecorator.java

```diff
@@ -0,0 +1,195 @@
+            cacheResolver = new MemoryCacheResolver();
+        }
+    }
+
+    @Override
+    public ResettableClassFileTransformer decorate(ResettableClassFileTransformer classFileTransformer) {
+        return new ResettableClassFileTransformer.WithDelegation(classFileTransformer) {
+
+            @Override
+            public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+                // load from cache
+                byte[] classCache = cacheResolver.getClassCache(loader, className);
+                if (classCache != null) {
+                    return classCache;
+                }
+
+                //transform class
+                classfileBuffer = classFileTransformer.transform(loader, className, classBeingRedefined, protectionDomain, classfileBuffer);
+
+                // save to cache
+                if (classfileBuffer != null) {
+                    cacheResolver.putClassCache(loader, className, classfileBuffer);
+                }
+
+                return classfileBuffer;
+            }
+        };
+    }
+
+    private static String getClassLoaderHash(ClassLoader loader) {
+        String classloader;
+        if (loader != null) {
+            classloader = Integer.toHexString(loader.hashCode());
+        } else {
+            //classloader is null for BootstrapClassLoader
+            classloader = "00000000";
+        }
+        return classloader;
+    }
+
+    interface ClassCacheResolver {
+
+        byte[] getClassCache(ClassLoader loader, String className);
+
+        void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer);
+    }
+
+    static class MemoryCacheResolver implements ClassCacheResolver {
+        // classloaderHashcode@className -> class bytes
+        private Map<String, byte[]> classCacheMap = new ConcurrentHashMap<String, byte[]>();
+
+        @Override
+        public byte[] getClassCache(ClassLoader loader, String className) {
+            String cacheKey = getCacheKey(loader, className);
+            return classCacheMap.get(cacheKey);
+        }
+
+        @Override
+        public void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer) {
+            String cacheKey = getCacheKey(loader, className);
+            classCacheMap.put(cacheKey, classfileBuffer);
+        }
+
+        private String getCacheKey(ClassLoader loader, String className) {
+            return getClassLoaderHash(loader) + "@" + className;
+        }
+    }
+
+    static class FileCacheResolver implements ClassCacheResolver {
+
+        private final File cacheDir;
+
+        FileCacheResolver(File cacheDir) {
+            this.cacheDir = cacheDir;
+            if (!cacheDir.exists()) {
+                cacheDir.mkdirs();
+            }
+
+            //clean cache dir on exit
+            FileUtils.deleteDirectoryOnExit(cacheDir);
```

Comment:
Deleting file shares the same issue of cache dir confliction.

Reply:
different agents will isolate their respective class cache.